src/bootstrap/builder.rs

```diff
@@ -284,18 +284,25 @@ impl StepDescription {
         }
     }
 
-    fn maybe_run(&self, builder: &Builder<'_>, pathsets: Vec<PathSet>) {
+    /// This might run the builder. Returns false if no targets can be determined indicating that
+    /// builder was not run
+    fn maybe_run(&self, builder: &Builder<'_>, pathsets: Vec<PathSet>) -> bool {
```

Comment:
Given that we return true if we did determine targets but they were excluded, I feel like the comment should also note that case - it otherwise also falls under "didn't run".

Reply:
I am having a hard time giving a nice description to this function. Instead of committing my current iteration, I will paste it here and then we can look at it?

```
    /// This might run the builder. If the any of the `pathsets` is excluded by `self.is_excluded`
    /// it will not run but will return true.
    /// If no target can be determined it will not run and will return false.
    /// Else it will run and return true.
    fn maybe_run(&self, builder: &Builder<'_>, pathsets: Vec<PathSet>) -> bool {
 ```

Reply:
At least we should adjust -- this might run `self`, not the Builder. Builder is essentially configuration, not something you run.

It seems a little odd that we return true if we are excluded -- that implies that e.g. `x.py test --host='' --exclude src/test/ui` would *not* cause the error this PR is trying to add, right? That doesn't seem logical to me.

Reply:
> It seems a little odd that we return true if we are excluded

I agree, but that is current behaviour. I would like to fix that in a future issue. Normally when I grab an issue I keep my fix contained to that issue. Would that plan of attack work for you?

But going forward I think we need to align on what is the most critical error - or should we emit both? If we both exclude the only path we are testing and give an empty host.

Reply:
> I agree, but that is current behaviour. I would like to fix that in a future issue. Normally when I grab an issue I keep my fix contained to that issue. Would that plan of attack work for you?

In general with bootstrap a targeted fix may not be the best one, because we don't want to add to confusion by adding checks that make little sense by themselves. This code isn't necessarily great and cleaning it up as we go is the main way we can make progress on changing that.

> But going forward I think we need to align on what is the most critical error - or should we emit both? If we both exclude the only path we are testing and give an empty host.

I think my confusion and maybe the miscommunication here is that "empty host" isn't the thing we're trying to give an error on -- at least not directly -- if that was the targeted fix being pursued, then a check in config.rs would be the right place. This PR is adding code interleaved with the decision of how to execute a specific part of the build (Step impl), and executes many times for every build.